File: example/1d-linear-convection/weno3/cpp/02d/src/flux/basic/RusanovFlux.cpp
#include "flux/basic/RusanovFlux.h"
#include <cmath>
#include <stdexcept>

namespace cfd {

Real RusanovFlux::compute_flux(Real uL, Real uR, Real wave_speed) const {
    Real F_L = wave_speed * uL;
    Real F_R = wave_speed * uR;
    Real Smax = std::abs(wave_speed);
    return 0.5 * (F_L + F_R) - 0.5 * Smax * (uR - uL);
}

void RusanovFlux::compute_flux_array(const Vector& uL, const Vector& uR, 
                                   Vector& flux, Real wave_speed) const {
    if (uL.size() != uR.size() || uL.size() != flux.size()) {
        throw std::invalid_argument("Flux array sizes do not match");
    }
    
    Real Smax = std::abs(wave_speed);
    
    for (size_t i = 0; i < flux.size(); ++i) {
        Real F_L = wave_speed * uL[i];
        Real F_R = wave_speed * uR[i];
        flux[i] = 0.5 * (F_L + F_R) - 0.5 * Smax * (uR[i] - uL[i]);
    }
}

} // namespace cfd

File: example/1d-linear-convection/weno3/fortran/cfd/04c/flux/rusanov_flux.f90
! ==================== rusanov_flux.f90 ====================
! Rusanov flux computation for OneFLOW-CFD
!
! Purpose: Compute numerical flux using Rusanov scheme
! Reference: V.V. Rusanov, "Calculation of Interaction of Non-steady
!            Shock Waves with Obstacles," J. Comp. Math. Phys. USSR, 1961
! Author: OneFLOW-CFD Team
! Date: Created
! =========================================================

module rusanov_flux_module
    use kinds, only: dp
    use config_module, only: CfdConfigType
    use mesh_module, only: MeshType
    implicit none
    
    private
    public :: rusanov_flux
    
contains
    
    ! Rusanov flux implementation
    subroutine rusanov_flux(q_face_left, q_face_right, flux, config, mesh)
        real(dp), intent(in) :: q_face_left(:), q_face_right(:)
        real(dp), intent(out) :: flux(:)
        type(CfdConfigType), intent(in) :: config
        type(MeshType), intent(in) :: mesh
        
        integer :: i
        real(dp) :: u_L, u_R, F_L, F_R, c_L, c_R, Smax
        
        c_L = config%wave_speed
        c_R = config%wave_speed
        
        do i = 1, mesh%nnodes
            u_L = q_face_left(i)
            u_R = q_face_right(i)
            F_L = c_L * u_L
            F_R = c_R * u_R
            Smax = max(abs(c_L), abs(c_R))
            flux(i) = 0.5_dp * (F_L + F_R) - 0.5_dp * Smax * (u_R - u_L)
        end do
    end subroutine rusanov_flux
    
end module rusanov_flux_module